doc/developer/design/20230119_ir_normalization.md

```diff
@@ -0,0 +1,147 @@
+
+Some of these skate close to optimizations.
+Turning an entire `Reduce` into a `Map` relies on the `group_key` being a unique key for the input collection.
+I think we should start judiciously here, especialy as the complexity can spiral.
+Again, we'll be running normalization alongside more sophisticated transforms that can perform more interesting optimizations.
+
+### Movement
+
+Many operators commute.
+When it is appropriate to do so, we should move commuting operators to canonical representations.
+For example, contiguous blocks of `Map`, `Filter`, and `Project` operators can be re-ordered to be `Map`s, `Filter`s, and then `Project`s.
+A tree of `Union` operators may benefit from having linear operators (as above, plus `Negate`, `FlatMap`) pushed down to their inputs, that multiple `Union`s can be merged.
+A tree of `Join` operators may benefit from having linear operators lifted to its root, that multiple `Join`s can be merged.
+
+Group-wise operators like `Reduce`, `Threshold`, `TopK`, etc. (and `ArrangeBy`) may need none of this.
+
+There are *optimizations* that allow us to move e.g. filters through the keys of reductions, but I think there may be some (debatable) value in doing structural and semantic normalization independently.
+Moving e.g. filters great distances may have value, but different transforms may disagree about the direction of movement (up or down) and we would like to keep normalization generally useful.
+
+WARNING: `Filter` operators do not introduce a specific evaluation order, and moving them is a recurring source of stress when it is used to guard erroring computation.
+We are "allowed" to commute it wherever we want, but we may need to push back on constructs that try to use `Filter` as an operator that "guarantees" much of anything, as it doesn't guarantee very much about the act of execution and evaluation, and once you panic (read: "introduce an record into the error output") you can't easily undo it.
+
+### Fusion
+
+Stacks of equivalent operators can often be consolidated into (at most) one operator.
+For example, each of `Map`, `Filter`, `Project`, `Union`, `Join`, `Negate`.
+Some forms of `Reduce` and `TopK` can be fused (`Distinct`, and same (key, order) respectively).
+We can restrict our attention to operators that can be fused without further investigation of the properties of their inputs.
+
+### Canonicalization
+
+Each operator should canonicalize its representation. 
+`Filter` should sort, deduplicate, and simplify its predicates.
+`Union` and `Join` should (could?) sort their inputs and permute their outputs.
+`Reduce` can sort its `group_key` and `aggregate` arguments.
+All operators should `reduce` their argument expressions to their simplest forms.
+
+Several of these proposals *introduce* permutations.
+These permutations are new AST nodes, but they should be easy to remove as part of a post-order traversal.
```

Comment:
If the permutation will be modeled by a `Project` node, maybe we can can just pull up all projections so that at the end (ideally) we only have one projection at the top that ensures that the end result is presented as requested to the client.

Reply:
Yes, I think there some "projection hoisting" that everyone can agree on. If nothing else "permutation hoisting" seems like an uncontroversial normalization. Perhaps there are "projection, but not permutation" moments that are harder to hoist, but .. permutations seemed uncontroversial to me.